src/services/query.js

```diff
@@ -0,0 +1,402 @@
+};
+
+
+/**
+ * Add multiple sources at once in the order they are given.
+ * @param {Array.<ngeox.QuerySource>} sources The sources to add to the query
+ *     service.
+ * @export
+ */
+ngeo.Query.prototype.addSources = function(sources) {
+  sources.forEach(this.addSource, this);
+};
+
+
+/**
+ * Clear the results.
+ * @export
+ */
+ngeo.Query.prototype.clear = function() {
+  this.clearResult_();
+};
+
+
+/**
+ * Issue a new request using a given map and a given object, which can be
+ * a coordinate or extent.
+ *
+ * NOTE: only coordinates are currently supported.
+ *
+ * @param {ol.Map} map The ol3 map object to fetch the layers from.
+ * @param {ol.Coordinate|ol.Extent} object The coordinate or extent to issue
+ *     the request with.
+ * @export
+ */
+ngeo.Query.prototype.issue = function(map, object) {
+
+  this.clearResult_();
+
+  if (object.length === 2) {
+    this.issueWMSGetFeatureInfoRequests_(map, object);
+  }
+};
+
+
+/**
+ * Issue a new WMS GetFeatureInfo request using a given map and a coordinate.
+ * For each visible layer of the map, if that layer has a source configured
+ * within this query service, then a query will be sent and the results
+ * will be stocked in the `ngeoQueryResult`.
+ *
+ * If multiple sources share the same url and use GML as info format, then
+ * only one request will be sent for all these sources.
+ *
+ * NOTE: Only GML info format are currently supported.
+ *
+ * @param {ol.Map} map The ol3 map object to fetch the layers from.
+ * @param {ol.Coordinate} coordinate The coordinate to issue the request with.
+ * @private
+ */
+ngeo.Query.prototype.issueWMSGetFeatureInfoRequests_ = function(
+    map, coordinate) {
+
+  var view = map.getView();
+  var projCode = view.getProjection().getCode();
+
+  var id;
+  var infoFormat;
+  var url;
+  var item;
+  var wmsGetFeatureInfoUrl;
+
+  var itemsByUrl =
+      /** @type {Object.<string, Array.<ngeo.QueryCacheItem>>} */ ({});
+
+  var resolution = /** @type {number} */ (view.getResolution());
+
+  map.getLayers().forEach(function(layer) {
+
+    // skip layers that are not visible
+    if (!layer.getVisible()) {
+      return;
+    }
+
+    // skip layers that don't have a source configured
+    id = this.getLayerSourceId_(layer);
+    if (!id || !this.cache_[id]) {
+      return;
+    }
+
+    item = this.cache_[id];
+    item['resultSource'].pending = true;
+    infoFormat = item.source.infoFormat;
+
+    // sources that use GML as info format are combined together if they
+    // share the same server url
+    if (infoFormat === ngeo.QueryInfoFormatType.GML) {
+      url = item.source.wmsSource.getUrl();
+      goog.asserts.assertString(url);
+      if (!itemsByUrl[url]) {
+        itemsByUrl[url] = [];
+      }
+      itemsByUrl[url].push(item);
+    } else {
+      // TODO - support other kinds of infoFormats
+      item['resultSource'].pending = false;
+    }
+
+  }, this);
+
+  goog.object.forEach(itemsByUrl, function(items) {
```

Comment:
itemsByUrl is not an array, it's an object.  Is there a native function for a forEach on an object ?